File: src/Euphoria.Graphics/Renderer.cs
using Euphoria.Math;
using SDL;
using static Euphoria.Graphics.SdlUtil;
using static SDL.SDL3;

namespace Euphoria.Graphics;

public unsafe class Renderer : IDisposable
{
    private readonly SDL_Window* _window;
    private readonly SDL_GPUDevice* _device;

    // TODO: This should be a UI renderer.
    public readonly TextureBatcher TextureBatcher;
    
    public Renderer(SDL_Window* window, in RendererInfo info)
    {
        _window = window;
        
        _device = Check(SDL_CreateGPUDevice(SDL_GPUShaderFormat.SDL_GPU_SHADERFORMAT_SPIRV, info.Debug, (byte*) null),
            "Create GPU device");

        Check(SDL_ClaimWindowForGPUDevice(_device, window), "Claim window for device");

        TextureBatcher = new TextureBatcher(_device);
    }
    
    public void Render()
    {
        SDL_GPUCommandBuffer* cb = Check(SDL_AcquireGPUCommandBuffer(_device), "Get command buffer");

        SDL_GPUTexture* swapchainTexture;
        uint width, height;
        Check(SDL_WaitAndAcquireGPUSwapchainTexture(cb, _window, &swapchainTexture, &width, &height),
            "Acquire swapchain texture");

        if (swapchainTexture == null)
            return;

        SDL_GPUColorTargetInfo targetInfo = new SDL_GPUColorTargetInfo()
        {
            texture = swapchainTexture,
            clear_color = new SDL_FColor() { r = 1.0f, g = 0.5f, b = 0.25f },
            load_op = SDL_GPULoadOp.SDL_GPU_LOADOP_CLEAR,
            store_op = SDL_GPUStoreOp.SDL_GPU_STOREOP_STORE
        };

        SDL_GPURenderPass* pass = Check(SDL_BeginGPURenderPass(cb, &targetInfo, 1, null), "Begin render pass");
        
        SDL_GPUViewport viewport = new SDL_GPUViewport()
        {
            x = 0,
            y = 0,
            w = width,
            h = height,
            min_depth = 0.0f,
            max_depth = 1.0f
        };
        SDL_SetGPUViewport(pass, &viewport);
        
        SDL_EndGPURenderPass(pass);
        
        TextureBatcher.RenderDrawList(cb, swapchainTexture, new Size<uint>(width, height));

        Check(SDL_SubmitGPUCommandBuffer(cb), "Submit command buffer");
    }

    public void Dispose()
    {
        TextureBatcher.Dispose();
        
        SDL_ReleaseWindowFromGPUDevice(_device, _window);
        SDL_DestroyGPUDevice(_device);
    }
}

File: src/Graphics/Renderer.cpp
// Copyright (c) Aquatic Games 2025. This file is licensed under the MIT license.
#include "Crimson/Graphics/Renderer.h"

#include <format>
#include <stdexcept>
#include <SDL3/SDL_gpu.h>
#include <SDL3_shadercross/SDL_shadercross.h>

#include "Crimson/Platform/Surface.h"
#include "Crimson/Util/Logger.h"
#include "Renderers/CanvasRenderer.h"

namespace Crimson
{
    static SDL_GPUDevice* _device;
    static SDL_Window* _window;

    static CanvasRenderer* _uiBatcher;

    void Renderer::Create()
    {
        const SDL_PropertiesID props = SDL_CreateProperties();
        SDL_SetBooleanProperty(props, SDL_PROP_GPU_DEVICE_CREATE_SHADERS_SPIRV_BOOLEAN, true);
#ifndef NDEBUG
        CS_DEBUG("Debugging enabled, enabling graphics debugging.");
        SDL_SetBooleanProperty(props, SDL_PROP_GPU_DEVICE_CREATE_DEBUGMODE_BOOLEAN, true);
        SDL_SetBooleanProperty(props, SDL_PROP_GPU_DEVICE_CREATE_PREFERLOWPOWER_BOOLEAN, true);
#endif

        CS_TRACE("Creating GPU device.");
        _device = SDL_CreateGPUDeviceWithProperties(props);
        if (!_device)
            CS_FATAL("Failed to create GPU device: {}", SDL_GetError());

        _window = static_cast<SDL_Window*>(Surface::GetHandle());
        SDL_ClaimWindowForGPUDevice(_device, _window);

        if (!SDL_ShaderCross_Init())
            CS_FATAL("Failed to initialize SDL_ShaderCross: {}", SDL_GetError());

        CS_TRACE("Creating UI batcher.");
        _uiBatcher = new CanvasRenderer(_device, SDL_GetGPUSwapchainTextureFormat(_device, _window));
    }

    void Renderer::Destroy()
    {
        delete _uiBatcher;
        SDL_ShaderCross_Quit();
        SDL_ReleaseWindowFromGPUDevice(_device, _window);
        SDL_DestroyGPUDevice(_device);
    }

    void Renderer::Present()
    {
        SDL_GPUCommandBuffer* cb = SDL_AcquireGPUCommandBuffer(_device);

        SDL_GPUTexture* swapchainTexture;
        SDL_WaitAndAcquireGPUSwapchainTexture(cb, _window, &swapchainTexture, nullptr, nullptr);

        if (!swapchainTexture)
            return;

        const SDL_GPUColorTargetInfo target
        {
            .texture = swapchainTexture,
            .clear_color = { 1.0f, 0.5f, 0.25f, 1.0f },
            .load_op = SDL_GPU_LOADOP_CLEAR,
            .store_op = SDL_GPU_STOREOP_STORE
        };

        SDL_GPURenderPass* pass = SDL_BeginGPURenderPass(cb, &target, 1, nullptr);
        SDL_EndGPURenderPass(pass);

        SDL_SubmitGPUCommandBuffer(cb);
    }
}
